delta-lake/common/src/main/scala/com/nvidia/spark/rapids/delta/GpuRapidsProcessDeltaMergeJoinExec.scala

```diff
@@ -0,0 +1,344 @@
+    output: Seq[Attribute],
+    outputExprs: Seq[GpuExpression],
+    targetRowHasNoMatch: GpuExpression,
+    sourceRowHasNoMatch: GpuExpression,
+    matchedConditions: Seq[GpuExpression],
+    matchedOutputs: Seq[Seq[Seq[GpuExpression]]],
+    notMatchedConditions: Seq[GpuExpression],
+    notMatchedOutputs: Seq[Seq[Seq[GpuExpression]]],
+    noopCopyOutput: Seq[GpuExpression],
+    deleteRowOutput: Seq[GpuExpression],
+    metrics: Map[String, GpuMetric])
+    extends Iterator[ColumnarBatch] with AutoCloseable with Arm {
+
+  Option(TaskContext.get()).foreach(_.addTaskCompletionListener[Unit](_ => close()))
+
+  private[this] val intermediateTypes: Array[DataType] = noopCopyOutput.map(_.dataType).toArray
+  private[this] var nextBatch: Option[ColumnarBatch] = None
+  private[this] val opTime = metrics.getOrElse(GpuMetric.OP_TIME, NoopMetric)
+  private[this] val numOutputRows = metrics.getOrElse(GpuMetric.NUM_OUTPUT_ROWS, NoopMetric)
+
+  override def hasNext: Boolean = {
+    nextBatch = nextBatch.orElse(processNextBatch())
+    nextBatch.isDefined
+  }
+
+  override def next(): ColumnarBatch = {
+    if (!hasNext) {
+      throw new NoSuchElementException("no more batches")
+    }
+    val result = nextBatch.get
+    nextBatch = None
+    result
+  }
+
+  override def close(): Unit = {
+    nextBatch.foreach(_.close())
+    nextBatch = None
+  }
+
+  private def processNextBatch(): Option[ColumnarBatch] = {
+    while (iter.hasNext) {
+        val input = iter.next()
+        val batch = withResource(new NvtxWithMetrics("process merge", NvtxColor.CYAN, opTime)) {
+          _ => processSingleBatch(input)
+        }
+      if (batch.numRows > 0) {
+        numOutputRows += batch.numRows
+        return Some(batch)
+      } else {
+        batch.close()
+      }
+    }
+    None
+  }
+
+  private def processSingleBatch(input: ColumnarBatch): ColumnarBatch = {
+    val (targetNoMatchBatch, targetMatchBatch) =
+      splitBatchAndClose(input, inputTypes, targetRowHasNoMatch)
+    val noopCopyBatch = closeOnExcept(targetMatchBatch) { _ =>
+      GpuProjectExec.projectAndClose(targetNoMatchBatch, noopCopyOutput, NoopMetric)
+    }
+    val bigTable = withResource(noopCopyBatch) { _ =>
+      val (sourceNoMatchBatch, sourceMatchBatch) =
+        splitBatchAndClose(targetMatchBatch, inputTypes, sourceRowHasNoMatch)
+      val sourceNotMatchedBatches = closeOnExcept(sourceMatchBatch) { _ =>
+        processProjectionSeries(sourceNoMatchBatch,
+          notMatchedConditions, notMatchedOutputs, deleteRowOutput)
+      }
+      withResource(sourceNotMatchedBatches) { _ =>
+        val sourceMatchedBatches = processProjectionSeries(sourceMatchBatch,
+          matchedConditions, matchedOutputs, noopCopyOutput)
+        withResource(sourceMatchedBatches) { _ =>
+          val allBatches = (noopCopyBatch +: sourceNotMatchedBatches) ++ sourceMatchedBatches
+          // annoyingly Table.concatenate does not gracefully handle the degenerate case
```

Comment:
Should we file an issue for this?

Reply:
I filed https://github.com/rapidsai/cudf/issues/12610